File: Python DSA/LinkedList/merge_two_sorted_lists.py
"""
MERGE TWO SORTED LISTS - Easy
Merge two sorted linked lists.

Logical Thinking:
1. Create dummy node
2. Compare values and append smaller
3. Handle remaining nodes

Logical Thinking:
def merge_two_lists(l1, l2):
    dummy = ListNode(0)
    curr = dummy
    while l1 and l2:
        if l1.val < l2.val:
            curr.next = l1
            l1 = l1.next
        else:
            curr.next = l2
            l2 = l2.next
        curr = curr.next
    curr.next = l1 or l2
    return dummy.next
"""

# TODO: Implement solution


# Test cases
if __name__ == "__main__":
    print(f"Testing Merge Two Sorted Lists...")
    print("Add your test cases here")

File: DSA/Linked List/Java/MergeTwoSortedLists.java
// LeetCode 21: Merge Two Sorted Lists
// You are given the heads of two sorted linked lists list1 and list2.

public class MergeTwoSortedLists {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        
        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                current.next = list1;
                list1 = list1.next;
            } else {
                current.next = list2;
                list2 = list2.next;
            }
            current = current.next;
        }
        
        // Append remaining nodes
        current.next = (list1 != null) ? list1 : list2;
        
        return dummy.next;
    }
    
    // Recursive approach
    public ListNode mergeTwoListsRecursive(ListNode list1, ListNode list2) {
        if (list1 == null) return list2;
        if (list2 == null) return list1;
        
        if (list1.val <= list2.val) {
            list1.next = mergeTwoListsRecursive(list1.next, list2);
            return list1;
        } else {
            list2.next = mergeTwoListsRecursive(list1, list2.next);
            return list2;
        }
    }
    
    public static void main(String[] args) {
        MergeTwoSortedLists solution = new MergeTwoSortedLists();
        
        // Create test lists: [1,2,4] and [1,3,4]
        ListNode list1 = new ListNode(1);
        list1.next = new ListNode(2);
        list1.next.next = new ListNode(4);
        
        ListNode list2 = new ListNode(1);
        list2.next = new ListNode(3);
        list2.next.next = new ListNode(4);
        
        ListNode merged = solution.mergeTwoLists(list1, list2);
        System.out.println("Two sorted lists merged successfully");
    }
}